src/core/dep-graph.ts

```diff
@@ -327,39 +308,60 @@ class DepGraphImpl implements types.DepGraphInternal {
     return node;
   }
 
-  private pathsFromNodeToRoot(nodeId: string): types.PkgInfo[][] {
+  private pathsFromNodeToRoot(
+    nodeId: string,
+    seenOnPath: Set<string>,
+  ): string[][] {
     const parentNodesIds = this.getNodeParentsNodeIds(nodeId);
-    if (parentNodesIds.length === 0) {
-      return [[this.getNodePkg(nodeId)]];
+
+    // No parents --> we're at root!
+    if (!parentNodesIds.length) {
+      return [[nodeId]];
     }
 
-    const allPaths: types.PkgInfo[][] = [];
-    parentNodesIds.map((id) => {
-      const out = this.pathsFromNodeToRoot(id).map((path) => {
-        return [this.getNodePkg(nodeId)].concat(path);
+    const allPaths: string[][] = [];
+    parentNodesIds
```

Comment:
same as below.. given how "hot" this method is, I think we'd benefit from a basic `for(const of)`